File: 0290-word-pattern/0290-word-pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        int n = pattern.length();
        List<String> str = new ArrayList<>();
        s += " ";
        StringBuilder sb = new StringBuilder();
        for(int i=0; i<s.length(); i++) {
            if(s.charAt(i) == ' ') {
                str.add(sb.toString());
                sb = new StringBuilder();
            }
            else {
                sb.append(s.charAt(i));
            }
        }
        if(str.size() != pattern.length()) return false;  
        Map<String, Character>mp1 = new HashMap<>();
        Map<Character, String>mp2 = new HashMap<>();
        for(int i=0; i<n; i++) {
            String curStr = str.get(i);
            char curCh = pattern.charAt(i);
            if(mp1.containsKey(curStr)) {
                if(mp1.get(curStr) != curCh) return false;
            }
            if(mp2.containsKey(curCh)) {
                if(!mp2.get(curCh).equals(curStr)) return false;
            }
            mp1.put(curStr, curCh);
            mp2.put(curCh, curStr);
        }
        return true;
    }
}

File: 0290-word-pattern/0290-word-pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        vector<string>str;
        s += " ";
        string temp = "";
        for(int i=0; i<s.size(); i++) {
            if(s[i] == ' ') {
                str.push_back(temp);
                temp = "";
            }
            else {
                temp += s[i];
            }
        }
        unordered_map<string, char>mp1;
        unordered_map<char, string>mp2;
        if(str.size() != pattern.size()) return false;
        for(int i=0; i<pattern.size(); i++) {
            if(mp1.find(str[i]) != mp1.end()) {
                if(mp1[str[i]] != pattern[i]) return false;
            }
            if(mp2.find(pattern[i]) != mp2.end()) {
                if(mp2[pattern[i]] != str[i]) return false;
            }
            mp1[str[i]] = pattern[i];
            mp2[pattern[i]] = str[i];
        }
        return true;
    }
};